pkg/serializer/jsonstream/compressor.go

```diff
@@ -0,0 +1,261 @@
+	}
+
+	c.input = inputBufferPool.Get().(*bytes.Buffer)
+	c.zipper = zlib.NewWriter(c.compressed)
+	n, err := c.zipper.Write(header)
+	c.uncompressedWritten += n
+
+	return c, err
+}
+
+// checkItemSize checks that the item can fit in a payload. Worst case is used to
+// determine the size of the item after compression meaning we could drop an item
+// that could actually fit after compression. That said it is probably impossible
+// to have a 2MB+ item that is valid for the backend.
+func (c *compressor) checkItemSize(data []byte) bool {
+	return len(data) < c.maxUnzippedItemSize && compression.CompressBound(len(data)) < c.maxZippedItemSize
+}
+
+// hasRoomForItem checks if the current payload has enough room to store the given item
+func (c *compressor) hasRoomForItem(item []byte) bool {
+	uncompressedDataSize := c.input.Len() + len(item)
+	if !c.firstItem {
+		uncompressedDataSize += len(jsonSeparator)
+	}
+	return compression.CompressBound(uncompressedDataSize) <= c.remainingSpace() && c.uncompressedWritten+uncompressedDataSize <= maxUncompressedSize
+}
+
+// pack flushes the temporary uncompressed buffer input to the compression writer
+func (c *compressor) pack() error {
+	expvarsTotalCycles.Add(1)
+	n, err := c.input.WriteTo(c.zipper)
+	if err != nil {
+		return err
+	}
+	c.uncompressedWritten += int(n)
+	c.zipper.Flush()
+	c.input.Reset()
+	return nil
+}
+
+// addItem will try to add the given item
+func (c *compressor) addItem(data []byte) error {
+	// check item size sanity
+	if !c.checkItemSize(data) {
+		return errTooBig
+	}
+	// check max repack cycles
+	if c.repacks >= maxRepacks {
+		return errPayloadFull
+	}
+
+	if !c.hasRoomForItem(data) {
+		if c.input.Len() == 0 {
+			return errPayloadFull
+		}
+		err := c.pack()
+		if err != nil {
+			return err
+		}
+		if !c.hasRoomForItem(data) {
+			return errPayloadFull
+		}
+		c.repacks++
+	}
+
+	// Write the separator between items
+	if c.firstItem {
+		c.firstItem = false
+	} else {
+		c.input.Write(jsonSeparator)
+	}
+
+	c.input.Write(data)
+	return nil
+}
+
+func (c *compressor) close() ([]byte, error) {
+	// Flush remaining uncompressed data
+	if c.input.Len() > 0 {
+		n, err := c.input.WriteTo(c.zipper)
+		c.uncompressedWritten += int(n)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Add json footer
+	n, err := c.zipper.Write(c.footer)
+	c.uncompressedWritten += int(n)
+	if err != nil {
+		return nil, err
+	}
+	// Add zlib footer and close
+	err = c.zipper.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	c.input.Reset()
+	inputBufferPool.Put(c.input)
+
+	expvarsTotalPayloads.Add(1)
+	expvarsBytesIn.Add(int64(c.uncompressedWritten))
+	expvarsBytesOut.Add(int64(c.compressed.Len()))
+
+	return c.compressed.Bytes(), nil
+}
+
+func (c *compressor) remainingSpace() int {
+	return maxPayloadSize - c.compressed.Len() - len(c.footer)
+}
+
+// Payloads serializes a metadata payload and sends it to the forwarder
```

Comment:
Is this a type or are we using SteamJSONMarshaler for the metadata payload ?

If we are using it: why ? It's not like the metadata payload can be split or anything.

Reply:
c/p from https://github.com/DataDog/datadog-agent/blob/33db6c888082da73e04a7d344b9c78ee3a72371d/pkg/serializer/split/split.go#L56-L57

Both have nothing to do with the metadata payload. Will edit the comments